File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.java
import java.io.*;
import java.util.*;


public class Main {
    
    public static int white = 0;
    public static int blue = 0;
    public static int[][] a;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());

        a = new int[n][n];

        StringTokenizer st;

        for(int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine(), " ");

            for(int j = 0; j < n; j++) {
                a[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        solution(0, 0, n);

        System.out.println(white);
        System.out.println(blue);

    }

    public static void solution(int row, int col, int size) {

        if(colorCheck(row, col, size)) {
            if(a[row][col] == 0) {
                white++;
            }
            else {
                blue++;
            }
            return;
        }

        int newboard = size / 2;
        solution(row, col, newboard);
        solution(row, col + newboard, newboard);
        solution(row + newboard, col, newboard);
        solution(row + newboard, col + newboard, newboard);	
    }
    
    public static boolean colorCheck(int row, int col, int size) {

        int color = a[row][col];

        for(int i = row; i < row + size; i++) {
            for(int j = col; j < col + size; j++) {
                if(a[i][j] != color) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.cc
#include <iostream>
#include <vector>

using namespace std;

int white = 0;
int blue = 0;
vector<vector<int>> v;

void sol(int row, int col, int size);
bool colorCheck(int row, int col, int size);

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    
    int N;
    cin >> N;
    
    v.resize(N, vector<int>(N));
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            cin >> v[i][j];
        }
    }
    
    sol(0, 0, N);
    
    cout << white << '\n';
    cout << blue << '\n';
    
    return 0;
}

void sol(int row, int col, int size) {
    if (colorCheck(row, col, size)) {
        if (v[row][col] == 0) white++;
        else blue++;
        return;
    }
    
    int newSize = size / 2;
    sol(row, col, newSize);
    sol(row + newSize, col, newSize);
    sol(row, col + newSize, newSize);
    sol(row + newSize, col + newSize, newSize);
}

bool colorCheck(int row, int col, int size) {
    int color = v[row][col];
    
    for (int i = row; i < row + size; i++) {
        for (int j = col; j < col + size; j++) {
            if (color != v[i][j]) {
                return false;
            }
        }
    }
    return true;
}